### PR TITLE
`Programming exercises`: Show build logs for submission results

### PR DESCRIPTION
**Pull request by laurenzfb**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

Currently, Artemis only shows the build log a submission result if the build has failed. Instructors should be able to view the build logs every time.

Addresses #8138 

### Description

In this PR, we save the build logs in the file system identified by the result ID. If the build log file for a submission result is available, it can be downloaded by clicking the 'View build logs' link. We further implement the functionality to regularly remove old build logs from the file system.

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Create a programming exercise
2. Once the template and solution build have finished, click on 'Show Template Submissions'
3. Verify that you see a 'View build log' link next to the result
4. Click on the link. It should download a log file
5. Verify the contents of the build log file are not empty

1. Participate in the Exercise and make a submission and wait for the result
2. Go to the participations overview and click on submissions for your participation
3. Verify that you see a 'View build log' link next to the result

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

image

## Summary by CodeRabbit

- **New Features**
	- Introduced the ability to save, retrieve, and delete build logs to/from files.
	- Added functionality to manage build log file expiration.
	- Implemented a feature to check for the availability of build log files for results.
	- Enhanced the user interface to display a link for viewing build logs if they are available.
- **Bug Fixes**
	- Adjusted the parameter passed to the `processNewProgrammingExerciseResult` method to ensure correct logic in processing programming exercise results.
- **Tests**
	- Added integration tests for build logs retrieval and storage functionality.
	- Expanded test coverage to include checks for logs availability for participation results.

### CODE REVIEW

File: .idea/runConfigurations/_template__of_Gradle.xml [hunk 1]
```diff
@@ -19,4 +19,4 @@
     <RunAsTest>false</RunAsTest>
     <method v="2" />
   </configuration>
-</component>
\ No newline at end of file
+</component>
```
Comment (rstief) [inline .idea/runConfigurations/_template__of_Gradle.xml:22]: This looks accidental :)

File: src/main/webapp/app/exercises/shared/participation-submission/participation-submission.component.ts [hunk 2]
```diff
@@ -91,39 +92,53 @@ export class ParticipationSubmissionComponent implements OnInit {
...
+
+                        this.isLoading = false;
+                    });
+                } else {
+                    // Get exercise for release and due dates
+                    this.exerciseService.find(params['exerciseId']).subscribe((exerciseResponse) => {
+                        this.exercise = exerciseResponse.body!;
+                        this.updateStatusBadgeColor();
+                    });
+                    this.fetchParticipationAndSubmissionsForStudent();
```
Comment (rstief) [inline src/main/webapp/app/exercises/shared/participation-submission/participation-submission.component.ts:139]: does the else-case also need a `this.isLoading = false`?

**Review by rstief — DISMISSED** [review]
Demonstrated in Testing Session
Code also looks good, two small comments :)

File: src/main/webapp/app/exercises/shared/participation-submission/participation-submission.component.ts (hunk 2 above)
Comment (laurenzfb) [reply src/main/webapp/app/exercises/shared/participation-submission/participation-submission.component.ts:139]: good point

**Review by milljoniaer — DISMISSED** [review]
code looks good 👍

**Review by rstief — DISMISSED** [review]
Thanks for addressing the comments :) lgtm

**Review by mateusmm01 — DISMISSED** [review]
code lgtm. also tested locally and works as expected

**Review by mateusmm01 — APPROVED** [review]

**Review by rstief — APPROVED** [review]
Reapprove after merge

**Review by milljoniaer — APPROVED** [review]

**Review by reschandreas — APPROVED** [review]
tested locally, and the logs are beautiful!

**Pull request merged**